### ISSUE TITLE
Dev Containers

### ISSUE THREAD

**Issue by abhibeckert** [zed-industries/zed#11473]
### Check for existing issues

- [X] Completed

### Describe the feature

A Dev Container is a system that runs your code and part of your development environment (Language Server, Unit Tests, Debugger, Build Scripts, etc) inside a container. Many people consider this a blocking feature, as it's arguably an industry best practice to manage projects with complicated dependencies.

Visual Studio Code has a great implementation of this feature. Essentially you describe your environment and dependencies in a `.devcontainer` directory filled with config files, and the editor spins up a container when you open the project - it only takes a few seconds and often needs. Your code and the majority of VSCode extensions will run inside the container.

This feature request has been previously discussed in #5347 Remote Development and is related to the `collab` feature, but it's not quite the same thing and should be tracked separately in the issue tracker.

**Comment by Moshyfawn** [comment]
Related #5347

**Comment by versecafe** [comment]
Getting started on this I'd love help from anyone who's interested, took a little bit to figure out GPUI, panel registration, and settings since they're not super documented but here's the branch, just got the basic UI for the panel up, starting on the actual logic using [bollard](https://crates.io/crates/bollard)

### UI Example *Very Early*
<img width="1402" alt="image" src="https://github.com/zed-industries/zed/assets/147033096/e0619ec5-f73e-4f58-b606-d0a3a29375ca">


### Branch
https://github.com/versecafe/zed/tree/dev-containers

@abhibeckert Any opinions on how closely you think it should match VS Codes implementation, both in UI and in features?

**Comment by ConradIrwin** [comment]
@versecafe Thanks for taking a pass at this! This is one of the top features for people in the remote development alpha too.

For remote servers we ended up choosing to use a modal instead of a panel because the primary use-case of showing the list is to select one and open a new workspace (it didn't seem useful enough to show the list permanently to take up screen real estate all the time). Are there other use-cases that mean it's helpful to have this in an always-open panel while working on code, or should we add this to the cmd-shift-o dialogue too?

As a heads up, @maxbrunsfeld has been exploring ideas for a worktree based remoting feature, which is probably a better starting point than the current remoting feature (which installs a full zed on the remote host and communicates over the internet).

What are the next steps on this in your mind?

**Comment by chloerei** [comment]
One of the features I rely on in vscode devcontainer is to enable extensions in the container. Since my host environment does not have dependencies installed, extensions may not work properly in the host environment.

For example, in devcontainer.json:

```
  "customizations": {
    "vscode": {
      "extensions": [
        "Shopify.ruby-lsp",
        "bradlc.vscode-tailwindcss"
      ]
    }
  }
```

**Comment by versecafe** [comment]
@chloerei Something like this maybe

`zed-devcontainer.json`
```json
{
  "zed": {
    "extensions": [
      { "git": "https://github.com/versecafe/zed-groovy" },
      { "zed": "kotlin" },
      { "zed": "base16" }
    ],
    "lsp": {
      "rust-analyzer": {
        "initialization_options": {
          "cargo": {
            "features": "all"
          }
        },
        "procMacro": {
          "ignored": ["server"]
        }
      }
    }
  }
}
```

Extensions both from Zed directly or direct git URL for custom ones + LSP overrides

**Comment by ConradIrwin** [comment]
It probably makes most sense to just auto-install any extensions on the remote that are installed on the local and which are needed for the current filetype. That way you (shouldn't) have to think about it in the common case.

**Comment by abhibeckert** [comment]
> It probably makes most sense to just auto-install any extensions on the remote that are installed on the local and which are needed for the current filetype. 

No definitely not. Some extensions, e.g. anything providing intellisense, consume a lot of resources (RAM, CPU, Disk I/O…) which might crash hard on a low end virtual server.

I’ve also seen node extensions rack up a few hundred dollars in usage costs by installing node extensions in VSCode a production server which would automatically be allocated additional RAM/CPU cores when load was high. Node is perfectly happy to consume ridiculous resources automatically especially if you load a production environment that might have gigabytes of compiled JavaScript files open the disk.

**Comment by lattice0** [comment]
devcontainers are the reason I'm not on zed yet. waiting 🙏

**Comment by samstride** [comment]
> It probably makes most sense to just auto-install any extensions on the remote that are installed on the local and which are needed for the current filetype. That way you (shouldn't) have to think about it in the common case.

Hmmm, if possible, it will be good to have a setting for auto-install local extension: on/off.

There are definitely situations where it will be good not to auto-install local extensions into devcontainer.

**Comment by maxbrunsfeld** [comment]
I think we can probably do something pretty targeted. In Zed, extensions expose a lot of static metadata about exactly what they provide. So we can just run extensions locally as normal, but when we detect for example that a particular language server extension needs to run on the remote host, we can copy just that one WASM binary over to the host, and run it there. We can probably upload it using the same SSH connection, so no need for the host to have internet access.

**Comment by lattice0** [comment]
devcontainers + web version would be AWESOME. Not even VS Code has remote web devcontainers. I miss that for development on my Galaxy Tab S9 Ultra

**Comment by BlueHatbRit** [comment]
As mentioned by @quackerex, it's an open specification now. There are already some tools which implement the spec which are editor agnostic, such as the [CLI reference implementation](https://github.com/devcontainers/cli). There's also [DevPod](https://github.com/loft-sh/devpod) which seems to have a GUI (I've not used it).

Those could be a way for someone to use Zed without giving up devcontainers for now. First class support would be incredible though, and the CLI project might be a good reference for the Zed implementation.

**Comment by forabi** [comment]
It would also be nice if Podman is supported out of the box. The VS Code implementation is pretty coupled to Docker and that comes with complexities when it comes to file permissions in rootless mode.

**Comment by lattice0** [comment]
I've been trying to use VSCode web on my android tablet and I have to say: basic things like mouse over errors won't work and VSCode team did not have interest to fix it. It's pretty broken. Also, it's been almost one year since I opened https://github.com/microsoft/vscode-remote-release/issues/9059 and they don't seem to want to support it, and their extension is closed source so we cannot change it.

Basically is Zed had: web version that works on tablets, self hosted version (but maybe also online tunnel support like vscode), devcontainer support, I'd switch immediately and forever.

Thanks!

**Comment by darltrash** [comment]
This is also necessary if you want to develop on an OS-Tree based system like Fedora Silverblue if not by running zed straight up inside toolbox, which afaik it doesn't actually work properly.

**Comment by imbev** [comment]
It would be ideal for Flatpak-packaged Zed to be able to use devcontainers with host podman.

**Comment by lattice0** [comment]
@andreytkachenko on Android, mouse on Sunshine+Moonlight is a very poor experience. I tried many solutions for vscode web version

**Comment by tidux** [comment]
> This is also necessary if you want to develop on an OS-Tree based system like Fedora Silverblue if not by running zed straight up inside toolbox, which afaik it doesn't actually work properly.

Yeah the solution for that is typically Distrobox which is... not great.  I actually switched my main Linux PC from Fedora Kinoite (Silverblue-KDE) to *Gentoo* because I was handbuilding so much of my dev envs anyway.

**Comment by nicraMarcin** [comment]
I'm also waiting for it.

**Comment by imbev** [comment]
> @chloerei Something like this maybe
> 
> `zed-devcontainer.json`
> 
> ```json
> {
>   "zed": {
>     "extensions": [
>       { "git": "https://github.com/versecafe/zed-groovy" },
>       { "zed": "kotlin" },
>       { "zed": "base16" }
>     ],
>     "lsp": {
>       "rust-analyzer": {
>         "initialization_options": {
>           "cargo": {
>             "features": "all"
>           }
>         },
>         "procMacro": {
>           "ignored": ["server"]
>         }
>       }
>     }
>   }
> }
> ```
> 
> Extensions both from Zed directly or direct git URL for custom ones + LSP overrides

Following the spec, I suggest the following:

```json
{
    "build": {
        "dockerfile": "Dockerfile"
    },
    "customizations": {
        "zed": {
            "extensions": [
                "git+https://github.com/versecafe/zed-groovy",
                "kotlin",
                "base16"
            ]
        }
    }
}
```

**Comment by haliliceylan** [comment]
> > @chloerei Something like this maybe
> > `zed-devcontainer.json`
> > ```json
> > {
> >   "zed": {
> >     "extensions": [
> >       { "git": "https://github.com/versecafe/zed-groovy" },
> >       { "zed": "kotlin" },
> >       { "zed": "base16" }
> >     ],
> >     "lsp": {
> >       "rust-analyzer": {
> >         "initialization_options": {
> >           "cargo": {
> >             "features": "all"
> >           }
> >         },
> >         "procMacro": {
> >           "ignored": ["server"]
> >         }
> >       }
> >     }
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Extensions both from Zed directly or direct git URL for custom ones + LSP overrides
> 
> Following the spec, I suggest the following:
> 
> ```json
> {
>     "build": {
>         "dockerfile": "Dockerfile"
>     },
>     "customizations": {
>         "zed": {
>             "extensions": [
>                 "git+https://github.com/versecafe/zed-groovy",
>                 "kotlin",
>                 "base16"
>             ]
>         }
>     }
> }
> ```

I highly recommend add version for extensions, which is not available vscode also causing some trouble.

**Comment by imbev** [comment]
> I highly recommend add version for extensions, which is not available vscode also causing some trouble.

In that case:
```
{
    "build": {
        "dockerfile": "Dockerfile"
    },
    "customizations": {
        "zed@v0": {
            "extensions": [
                "git+https://github.com/versecafe/zed-groovy",
                "kotlin",
                "base16"
            ]
        }
    }
}
```

**Comment by mathroc** [comment]
the spec says :

> Tool-specific configuration. Each tool should use a JSON object subproperty with a unique name to group its customizations.

so it can be whatever works for Zed in `customizations.zed`

**Comment by gintsmurans** [comment]
Just to add my two cents as I am also waiting for this in zed. This is how we are using `devcontainer.json`:

```
{
    "name": "External Services",
    "dockerComposeFile": [
        "../docker-compose.yml"
    ],
    "service": "develop",
    "runServices": [
        "develop"
    ],
    "shutdownAction": "none",
    "workspaceFolder": "/srv/app",
    "customizations": {
        "vscode": {
            "settings": {
                "python.defaultInterpreterPath": "/srv/venv/bin/python3",
                "python.languageServer": "Pylance",
                "python.analysis.typeCheckingMode": "strict",
                "python.analysis.autoImportCompletions": true,
                "[python]": {
                    "editor.formatOnSave": true,
                    "editor.defaultFormatter": "ms-python.black-formatter"
                },
                "python.analysis.extraPaths": [
                    "submodules/py-shared-lib"
                ],
                "python.autoComplete.extraPaths": [
                    "submodules/py-shared-lib"
                ]
            },
            "extensions": [
                "ms-python.python", // Python
                "ms-python.vscode-pylance", // Python autocomplete
                "ms-python.black-formatter", // Python formatter
                "github.copilot", // Github co-pilot
                "aaron-bond.better-comments" // To highlight some comments
            ]
        }
    }
}

```

**Comment by pascalbreuninger** [comment]
Disclosure: I'm a maintainer of DevPod

[DevPod](https://github.com/loft-sh/devpod/releases/tag/v0.6.0) just added support for Zed with the latest release. This doesn't apply zed specific settings like extensions and configuration to the workspace but it let's you build and run your devcontainers and then connect to them from Zed

**Comment by gintsmurans** [comment]
> Disclaimer: I'm a maintainer of DevPod
> 
> [DevPod](https://github.com/loft-sh/devpod/releases/tag/v0.6.0) just added support for Zed with the latest release. This doesn't apply zed specific settings like extensions and configuration to the workspace but it let's you build and run your devcontainers and then connect to them from Zed

Oh, this is interesting, will give it a try.

**Comment by ConradIrwin** [comment]
Thanks @pascalbreuninger!

For the rest of dev containers, configuration is still something I want to figure out – Zed's approach to extensions means it's relatively cheap to run whatever you have installed in the container, but I can see an argument for putting it in "vanilla mode" if you really don't want any language servers running.

**Comment by francoism90** [comment]
@ConradIrwin It's pain when using something like Flatpak.

**Comment by ConradIrwin** [comment]
What is?

**Comment by teohhanhui** [comment]
For those who want a first-party solution, there's https://github.com/devcontainers/cli

It doesn't support all the features, but it seems to work well enough for basic usage. 

Example:

    devcontainer up --workspace-folder .

**Comment by mathroc** [comment]
> Devcontainer itself is really just some nice workflow stuff on top of Docker containers. So maybe the right first Zed feature is remote development workflow for containers in general, like SSH. Devcontainer support probably adds on to that.

You right about that, the first step to have "native" devcontainer support is probably to have support to run the remote server inside a docker container. 

> I think the key thing needed from Zed to support devcontainer is how to run the LSP server from within the dev container. 

That part is probably nothing to worry about, once Zed is running with a server in the container, like it does with SSH everything will work the same.

---

In fact you can already use Zed and it's remote dev capabilties with docker container, the downside is that you need to run an ssh server in your container and connect to it using SSH, it's more setup than needed but once everything is in place the experience is great. You can do that by adding the `"ghcr.io/devcontainers/features/sshd:1": {}` feature to your devcontainer.json config to install sshd in your container. Then to configure and run it I'm using devpod, it's in fact transparent, I just have to run `devpod up . --ide zed` and it'll start the container(s), start sshd in the main container, start zed, and connect it via ssh with the appropriate user / key to that container. 

I think the next steps should be:

1. support connecting zed to docker containers the same way we can with ssh (either from the UI or from the CLI when starting zed). It will allow tools like devpod will be able to simplify the boot procedure, or to connect manually more easily
2. have a devcontainer UI in zed to start a devcontainer project, monitor the containers, etc. Bonus if starting the devcontainer project from tools like devpod is compatible with the future zed UI (ie: zed understands that it was starting in "devcontainer mode", not just connected to a random docker container)

**Comment by goncalossilva** [comment]
This is also limiting on a recurring basis on my/our end.

We don't actually use dev containers, just plain containers and a tiny orchestrator, but if Zed follows [the plan above](https://github.com/zed-industries/zed/issues/11473#issuecomment-2576470869):

> 1. support connecting zed to docker containers the same way we can with ssh

This would address our use case too, since we just need a way to connect to a locally running container.

**Comment by LockedThread** [comment]
Also looking forward to this.

**Comment by sp6370** [comment]
Eagerly waiting for this to land!!!!

**Comment by imbev** [comment]
When implementing this, please attempt compatibility with both Docker and Podman. VSCode's devcontainer integration is inconsistent when used with Podman.

**Comment by rgov** [comment]
> When implementing this, please attempt compatibility with both Docker and Podman. VSCode's devcontainer integration is inconsistent when used with Podman.

I would [report bugs to Podman](https://github.com/containers/podman/issues) for any incompatibilities (both with Zed and with VSCode), since they boastfully advertise that you can "just `alias docker=podman`" and that it faithfully implements the Docker Engine API. As long as this feature uses the API correctly, it's not Zed's responsibility to work around Podman bugs.

**Comment by forabi** [comment]
I would suggest focusing on making sure both rootless and rootful containers are supported rather than specifically optimizing for a specific engine.

**Comment by francoism90** [comment]
@rgov With Podman, you could also use a different devcontainer image or init flags.

This means you simply run the services with systemd, and don't do any port publishing in the editor itself. For instance, you run vite outside the editor.. this also helps a lot when your editor crashes or is slow.

I'm not saying it's the best approach (think this is unsupported). Some may want to actually do everything inside the editor. I'm kinda mixed about that.

**Comment by Muthukamalan** [comment]
Hi Everyone, 

is that feature shipped?

**Comment by versecafe** [comment]
@Muthukamalan it half exists between remoting configs and tasks you can effectively make a dev container with 5-10 lines of configs but there's not a proper UI and menu for spinning them up yet

**Comment by Muthukamalan** [comment]
Thanks @versecafe for prompt response.

following was my devcontainer.json file in vscode. could you guide what are all config need to change to adapt zed.
this there any blog/ article I can follow for the setup

```json

{
	"name": "Go",
	"image": "mcr.microsoft.com/devcontainers/go",
	"features": {},
	"postCreateCommand": "go version",
	"remoteUser": "root",
	"customizations": {
    	"vscode": {
        	"extensions":[
				"golang.go",
				"eamodio.gitlens"
			]
         	}
	}
}

```

**Comment by airtonix** [comment]
please don't create a separate devcontainer file spec. [follow the standard](https://containers.dev/implementors/spec/).

```json
{
	"name": "Go",
	"image": "mcr.microsoft.com/devcontainers/go",
	"features": {},
	"postCreateCommand": "go version",
	"remoteUser": "root",
	"customizations": {
    	"zed": {
        	"extensions":[
              { "git": "https://github.com/versecafe/zed-groovy" },
              { "zed": "go" },
			]
         }
	}
}
```

**Comment by airtonix** [comment]
> could you guide what are all config need to change to adapt zed. this there any blog/ article I can follow for the setup

There isn't (as far as I know), but i imagine you'll want to : 


1. follow the clues @mathroc [provides](https://github.com/zed-industries/zed/issues/11473#issuecomment-2576470869)
2. use a `.devcontainer/devcontainer.json` [like i mention](https://github.com/zed-industries/zed/issues/11473#issuecomment-2861030875)
3. you'll need to follow the instructions as outlined by the [devcontainers/features/sshd](https://github.com/devcontainers/features/blob/main/src/sshd/README.md) repo, which means:
  - annotating your `.devcontainers/docker-compose.yml` so that `- "2222:2222"` is listed under `ports`
  - run some kind of script in `postCreateCommand` that sets a password for the user your container is using (what user that is 🤷🏻 )


All of this is easier of course when you're 

> generally awesome at docker

**Comment by SemenMartynov** [comment]
Subscribe to updates on this issue.
For me, this is the main reason why I use vscode now.

**Comment by p0fi** [comment]
On top of docker or podman zed should also look at apple containers of macOS 26.

**Comment by AskeyNil** [comment]
Also looking forward to this.

**Comment by nanwangnongfu** [comment]
mark

**Comment by versecafe** [comment]
@p0fi it should be possible to make the actual container running doable for any OCI compatible runner such as apple containers the only issue is how many different runners are shipped as part of zed itself unless the setup becomes implementable through extensions with docker as the default and then options like orbstack or podman to build and run

**Comment by wold9168** [comment]
This work needs to be scheduled. 👍

**Comment by LockedThread** [comment]
I spoke to the Zed folks at RustConf and they have this on their radar. Looking forward to its implementation.

**Comment by airtonix** [comment]
> On top of docker or podman zed should also look at apple containers of macOS 26.

@p0fi Aren't apple-containers just a vm running linux, and therefore normal docker? 

You can already use Lima or Podman Desktop as far as I know. Nothing revolutionary here that can't be solved with : 

```sh
mymac> docker context create mydockerhostvm --docker "host=ssh://theapplecontainervmipaddress"                                                               
# (ssh/tcp whatever...)
```

or even just using `DOCKER_HOST=somethingsomethingsomethingdarkside`

edit. yep. nothing revolutionary https://github.com/apple/container/discussions/106

**Comment by oWretch** [comment]
Apple containers are just standard containers at runtime, but the challenge is it doesn't have a docker-compatible API (i.e. docker.sock) for interaction. See apple/container#66, where the docker API compatibility is closed as not planned.

**Comment by shide1989** [comment]
I really hope they can create a first version of this feature very soon. Like me, I guess some folks will want to get out the MS VSCode editor to access containers

**Comment by nick-walt** [comment]
A lot of great technology suggestions in this thread but not much articulation of the use cases, related problems to solve and desirable user experiences in the solutions. 

There seems to be at least two primary use cases or requirements for remote or split development:
 - run compute intensive workloads on a headless backend such as a remote server
 - run untrusted workloads in a fully sandboxed environment.

Both use cases can be desired and used at the same time. Arguably being able to offload as much compute intensive workload to a remote backend inherently supports a high degree of sandboxing.

It could also be desirable to enable full application/configuration of Zero Trust security protocols on a containerised dev environment should the user request it.

The technology might determine the degree of sandboxing and overall security, as well as the usability or ease of configuration. Docker could be considered the least secure sandbox and LXC system containers the most secure, with well implemented rootless OCI containerisation somewhere in between. 

QEMU-KVM hypervisor technology might be off the table but realistically does it have to be? Is it really in the too hard basket? 

Probably the most common use case is to be able to run Zed client on a host with full, direct Vulcan GPU acceleration and essentially have the entire dev environment isolated in a secure, fully sandboxed containerised Zed server, where the client is as lightweight and separate as the user desires. A container that is then accessed directly through Linux Sockets, or via SSH or some other network protocol — depending on whether the isolated environment is local or remote. 

Another example for local and remote workload isolation is to fully sandbox the AI, whether the AI is a cloud service or self hosted. The requirement is that the user is not obliged to run any AI on the client on the host. 

Talking about container technologies — is there any reason why Incus LXC and Incus OCI could not be supported along with Docker and OCI/Podman?

**Comment by teohhanhui** [comment]
@nick-walt This feature request is about [Dev Containers](https://containers.dev/). I think you should open a separate issue for your use case (or maybe there are other existing issues), as it doesn't seem to be the same at all.

**Comment by nick-walt** [comment]
Ahh, ok so this thread isn't a discussion about the generic development container concept which the OP seemed to be about. Yeah, why not stick to the specification? 😆👍

**Comment by rihards-simanovics** [comment]
Devcontainer as written in Microsoft [spec](https://containers.dev/implementors/spec/) is very powerful especially for quick local environment spin-up and consistency between developer IDE config, and maybe even some level of security as it is containerised inside docker container with only a few poked holes for basic inter-op with host system.

One thing it doesn't really allow you to do, at least not easily on Linux, is develop GUI apps as it needs direct access to the display compositor, not sure if it's a problem in Windows or macOS.

**Comment by probitcarwyn** [comment]
Something to keep in mind when looking at containers on linux (including devcontainers) is that Docker Desktop (as opposed to Docker Engine) on Linux uses a VM just like on macOS and Windows. See the following for their rationale:

- [Why does Docker Desktop for Linux run a VM?](https://docs.docker.com/desktop/troubleshoot-and-support/faqs/linuxfaqs/#why-does-docker-desktop-for-linux-run-a-vm)

Worth comparing to the Docker Engine notes on security:

- [Docker Engine security](https://docs.docker.com/engine/security/)

The short version is that if you want to avoid the overhead of the VM on Linux, there are a bunch of security related things to consider, several of which are not the default for Docker Engine. Most Linux devs using Docker are likely using Engine, not Desktop, and most I've encountered have been oblivious to some or all of these security considerations :(

I mention this as it's very much related to why some people (including me) are interested in podman in this context as their choices with these defaults (user namespaces, rootless, daemonless) are a better story when you're trying to avoid the VM.

Unfortunately it's still a battle getting this stuff to play ball in vscode too.

There's an opportunity to do this better here (still using the devcontainer spec).

**Comment by rihards-simanovics** [comment]
> Something to keep in mind when looking at containers on linux (including devcontainers) is that Docker Desktop (as opposed to Docker Engine) on Linux uses a VM just like on macOS and Windows.

That's an interesting thing I didn't know about.

Just tested to see when running VSCode devcontainer, I think it's not created in qemu (VM) as it accesses the docker engine directly. At least I think it is so, correct me if I'm testing it wrong, but had a look in this stack overflow answer: https://stackoverflow.com/a/71643762.

I am running Ubuntu 25.04 not sure if this is the reason why, was always under the impression that docker desktop deb package was simply a wrapper for the docker engine. Looks like I'll need to do some more digging.

**Comment by nick-walt** [comment]
> One thing it doesn't really allow you to do, at least not easily on Linux, is develop GUI apps as it needs direct access to the display compositor, not sure if it's a problem in Windows or macOS.

It seems like secure integration between a host and containerised or virtual machine based guests is a problem being overlooked. Right now if we want to spin up a virtual machine, lxc or oci container on Linux there is essentially no well defined pathway to achieving GPU acceleration of locally running guests using the host's GPU. Want snappy RDP connections to those local guests? Forget it. What about Venus for Vulcan acceleration on the GPU? Still an alpha level implementation requiring much hoopla even through Kernel version 6.13 and above supports it, as does QEMU 9.20 (or thereabouts). And the most promising technology, Virtio-GPU Native Context, is just vapourware.

So if we want secure GUI workload isolation in strong sandboxes while experiencing a usable level of integration with the host with good performance there is nothing right now. To run GUI apps in an OCI container right now we have to punch privileged holes through the container essentially making the endeavor a pointless exercise.

It is almost bizarre that the two biggest Linux DE creators, KDE and Gnome, are not working with QEMU, OCI and LXC on integrating sandboxed environments into their new Wayland Desktop Environments.

**Comment by rihards-simanovics** [comment]
Agree with all above.

> It is almost bizarre that the two biggest Linux DE creators, KDE and Gnome, are not working with QEMU, OCI and LXC on integrating sandboxed environments into their new Wayland Desktop Environments.

Do they have any issues with this open in their respective repositories? I, like most people, follow Gnome and KDE development on a very surface level, so I am unaware of the developments in that area until they are basically released.

Perhaps, like most of the time, they are likely waiting for community feedback, or fighting about ways to implement it apu... so we could make some noise there, so to speak 😅.

**Comment by mathroc** [comment]
guys, interesting topics but we're getting off track here ;)

ps: maybe look at flatpak for sandboxed apps

**Comment by rihards-simanovics** [comment]
> guys, interesting topics but we're getting off track here ;)

This is not exactly off-topic, as in my case, the issue involves the inability to **develop** a GUI application in a devcontainer on Linux, which I hope, when the feature is developed, is at least considered. I don't want to spin up a whole VM to create a Qt application using Python, or, as Nick-Walt pointed out, to do "secure GUI workload isolation".

If there is something in particular you find to be off-topic, please specify.

> ps: maybe look at flatpak for sandboxed apps

I would like to gently point out that you have missed the point. Running a web server inside a container is a reasonably simple endeavour, as you only need to open a couple of ports and allow some access via SSH. Still, when it comes to any graphics processing or where the container needs to have a (preferably) secure access to the GPU, things fall apart, as Nick-Walt explained. Whether it's a mobile native app or a desktop application, the lack of host-client APIs makes it impossible to develop securely without exposing the client to the host's APIs, thereby undermining the argument that a devcontainer provides a secure development environment.

If I misquoted or misunderstood something, please let me know.

**Comment by teohhanhui** [comment]
> This is not exactly off-topic, as in my case, the issue involves the inability to develop a GUI application in a devcontainer on Linux, which I hope, when the feature is developed, is at least considered.

Again, that's not what this issue is about, hence it's off topic. Perhaps you should open an issue on [devcontainers/spec](https://github.com/devcontainers/spec) repo?

EDIT: Seems like there's https://github.com/devcontainers/features/tree/eea561a9ad45c383c2d9832d2da031410be8b5a0/src/desktop-lite

**Comment by rihards-simanovics** [comment]
I see, fair point, spec might be a better place, thanks for the link!

**Comment by abhibeckert** [comment]
> the inability to develop a GUI application in a devcontainer on Linux, which I hope, when the feature is developed, is at least considered

To be honest I’m not sure there’s a huge demand for that.

Dev Containers are mostly necessary for server side software development where your code often needs a complex environment involving multiple servers to run. Some of the containers I use in VS Code create a dozen virtual machines with different operating systems and totally different software (often the same software, but different versions or configurations due to dependency conflicts).

You might, for example, want a dozen instances of Apache and Ngnx all listening to traffic on network port 80… you need them all to have their own IP address for that to be possible.

Creating that setup manually without dev containers can take weeks of work and if you ask a dozen people on the team to do that, they’ll all do things slightly differently and none of them will really match your production environment. The worst inefficiency is code will be written that might only work on the author’s personal laptop.

That’s the problem this solves. A dev container specifies and builds a test environment for code to be executed in, which can be created in minutes rather than weeks, is exactly the same for everyone in the team, and can be changed/updated pretty much in lockstep with changes to the production server environment.

It could be useful for GUI software, but its orders of magnitudes less important. This is really about software that runs on servers.

**Comment by blue42u** [comment]
> > the inability to develop a GUI application in a devcontainer on Linux, which I hope, when the feature is developed, is at least considered
>
> To be honest I’m not sure there’s a huge demand for that. [...] It could be useful for GUI software, but its orders of magnitudes less important. This is really about software that runs on servers.

It's worth noting that VSCode has an option to mount the local Wayland socket inside the container (`@dev.containers.mountWaylandSocket`). I use this option regularly to develop my GUI application, i.e. in a single container that provides the common dependencies and development tools. I would use such a feature if it was available in Zed, although I agree it is a lower priority than basic devcontainer support.

**Comment by nick-walt** [comment]
Will follow up on some links for the container and sandboxing technology underlying the Dev Container specification which I imagine are indeed unrelated. 

One question about the DC spec — if Zed wanted to adopt and add to the spec would they be able to contribute to the project and collaborate with the project owners with minimal friction? Might be a naive question.

**Comment by NewtonChutney** [comment]
> It could be useful for GUI software, but its orders of magnitudes less important. This is really about software that runs on servers.

Yeah, especially on non-linux OSes.. when there's no widespread container implementation.. 
Windows has containers, but only if you pay them for the pro license
I'm not sure about macos

**Comment by airtonix** [comment]
> It could be useful for GUI software, but its orders of magnitudes less important. This is really about software that runs on servers.

That' not really for you to decide though is it? 

It's also a super valid use case. one that is specifically  supported by toolbox: 

> https://containertoolbx.org/use/#desktop-applications

**Comment by psiphi75** [comment]
> Dev Containers are mostly necessary for server side software development where your code often needs a complex environment involving multiple servers to run. Some of the containers I use in VS Code create a dozen virtual machines with different operating systems and totally different software (often the same software, but different versions or configurations due to dependency conflicts).
> 

That’s what you’ve used it for. Doesn’t mean that everyone uses devcontainers like that. The way our team use it is for a repeatable build environment for developing, deploying and debugging embedded applications. VS Code has nailed this. Using VS Code and devcontainers I can write c and cpp code, deploy it to an embedded device and debug it too. I can install developer tools (clang format, gcc, etc) inside the container and I know everyone in my team will be running the exact same setup. So I just commit tooling inside the devcontainer config and the whole team gets it. 

Another use for devcontainers is to run our GUI app. The GUI app simulates the embedded UI and makes prototyping much quicker.

Everyone in the team works from home on occasion, or other remote locations (like the other side of the world). Our laptops have windows and VS Code makes it seamless to remote into the devcontainer and work remotely.

Spinning up a new team members with the right tools takes no time as well. Install Docker, VS Code and git clone the repo and you’re good to go.

All that being said, I’d love to move away from VS Code.

**Comment by rihards-simanovics** [comment]
All I have to say is that I hope they at least consider GUI & embedded development in the future, even if not right away. There are numerous toolchains available for web developers, but fewer options are available for GUI or embedded development - at least outside of Windows. A bonus for me would be if it worked well on Linux; my workstation can run Windows 11 quite easily and has in the past, but the OS is so bloated that the debugger would chug every 10 minutes. In contrast, on Linux, it runs smoothly.

I worked in a team not long ago, and the ability to set up the environment in minutes is almost magical, as I recall, in the days prior, I would have had to set out a full day to set everything up. Now I can format my hard drive, install a new OS, and be back to work in just a few clicks, all the way down to an exact setting in the IDE.

Also, being run in Docker, it would help tremendously for the environment to be isolated from he host system. I've read a few alarming articles about various packages on npm being infected with malware, as well as those on pip.

Having to run VMware with Windows 11 each time I need the actual job done just doesn't cut it anymore for me.

**Comment by mateka** [comment]
> > It could be useful for GUI software, but its orders of magnitudes less important. This is really about software that runs on servers.
> 
> Yeah, especially on non-linux OSes.. when there's no widespread container implementation.. Windows has containers, but only if you pay them for the pro license I'm not sure about macos

There are independent project implementing containers for MacOS, eg. https://github.com/lima-vm/lima or https://podman.io. In the newest macOS Apple started creating it's own container framework: https://github.com/apple/container, which of course is a little different than Docker ;)

**Comment by mathroc** [comment]
Hi! 

> This is not exactly off-topic, as in my case, the issue involves the inability to develop a GUI application in a devcontainer

I did missed that part, but still, that's the responsibility of the container runtime, not even of the devcontainer spec, nor how Zed implements its integration of devcontainers. And as far as I know, you should be able to do whatever you want with devcontainers, if it's supported by docker / docker-compose or your container runtime it should work just fine when it's integrated with Zed

>  on Linux, which I hope, when the feature is developed, is at least considered

It's not like Zed would even have the option to not support it. Zed will likely only support docker / docker-compose at first, so you'll be covered to develop UI apps. And you can already use devcontainers with Zed via devpod as I mentionned [here](https://github.com/zed-industries/zed/issues/11473#issuecomment-2576470869). there will be some manual configuration to do but you can forward the x11/wayland socket inside your container, you can expose gpu if you wish to, etc.

that's why I think the discussion is getting off track. As far as I understand this issue is about :

* Zed connecting directly to local or remote containers using docker or another container runtime. 
* managing those containers following the devcontainer spec

whatever else is executed on the containers is something different. maybe Zed can provide shortcuts to configure wayland/x11 socket forwarding to the container, but that's outside of the spec afaict and not blocking that would only be a shortcut to something that is _already_ possible. (same thing with auto forwarding of ports for network app development)

**Comment by gnarea** [comment]
Folks, please stay on topic. If you're requesting something that's not part of the dev containers spec, then you've gone off topic. 

Hundreds of us have reacted positively this issue, so that's potentially hundreds of us who are getting notifications for something we didn't sign up for.

**Comment by rihards-simanovics** [comment]
> It's not like Zed would even have the option to not support it. Zed will likely only support docker / docker-compose at first, so you'll be covered to develop UI apps. And you can already use devcontainers with Zed via devpod as I mentionned [here](https://github.com/zed-industries/zed/issues/11473#issuecomment-2576470869). there will be some manual configuration to do but you can forward the x11/wayland socket inside your container, you can expose gpu if you wish to, etc.

Hi @mathroc, thanks for the reply. That's all I'm really concerned about. Whether it comes in initially or later doesn't really bother me much. For now, I do have a setup that holds us together with PVA glue, but it will do for now 😅. Thanks for the links as well. I'll research those later.

> Hundreds of us have reacted positively this issue, so that's potentially hundreds of us who are getting notifications for something we didn't sign up for.

@gnarea, you are welcome to customise the notification settings for this issue or set up an email filter; It is not my responsibility to keep your mailbox clutter-free.

**Comment by ggjulio** [comment]
@rihards-simanovics maybe creating a GH discussion would be a good idea. Issues are not great for discussing. 😄
That way no one gets spammed and people who have questions/suggestions, or want to go a little out of topic can do it without notifying everyone following this issue ? 

https://github.com/zed-industries/zed/discussions

**Comment by rihards-simanovics** [comment]
@ggjulio, there is no need; I already said what I needed to say. Dislike my comment all you like, but the reality is, as I already said, it's not my responsibility to keep your mailbox free of clutter.

I know issues are not great for discussions. I do a triaging job on the OS project all the time. Reality is, I'm asking for something, not having a discussion. What is a discussion for one is a genuine problem for another. I am happy to create a discussion, but I already added my 5 cents, so I'm quite content with where this is.

Edit:
Not sure if people would want to continue the "discussion" but here is the discussion I've made jic: https://github.com/zed-industries/zed/discussions/41765

**Comment by KyleBarton** [comment]
Hello, if anyone feels comfortable sharing a quick screen-record snippet of how they open/configure/use their dev container environments with another tool like VS Code, it would be super helpful. It's important to get the UX right, so seeing real-world usage would give us more data as we dig further in. Thanks!

**Comment by jtkiley** [comment]
> Hello, if anyone feels comfortable sharing a quick screen-record snippet of how they open/configure/use their dev container environments with another tool like VS Code, it would be super helpful. It's important to get the UX right, so seeing real-world usage would give us more data as we dig further in. Thanks!

Are you looking for a really quick snippet of clicking through things (like creating a devcontainer) or a couple minutes narrated or something else?

I'll pull something together when I have a few minutes.

**Comment by KyleBarton** [comment]
@jtkiley any amount of detail you're willing to provide would be valuable -- appreciate you taking the time.

**Comment by oWretch** [comment]
As a starting point, Microsoft's documentation on using Dev Containers in VS Code is a great starting point (and probably the reference implementation).

[Developing inside a Container](https://code.visualstudio.com/docs/devcontainers/containers) is the starting point, and has a couple of quick start examples at the beginning before delving into all the settings and customisation that is available.

[Dev Containers tutorial](https://code.visualstudio.com/docs/devcontainers/tutorial) is a great step-by-step guide on how to do the basics of a container, and will give you a great introduction to the experience.

[Beginner's Series to: Dev Containers](https://learn.microsoft.com/en-us/shows/beginners-series-to-dev-containers/?wt.mc_id=MVP_462083) is a Microsoft produced video series (8x ~5 min videos) that steps you through Dev Containers visually.

**Comment by valentin-p** [comment]
Here are some very informative videos.

https://youtu.be/Hx-GDw4XfOQ?si=B1dFyvq83NCQDSNN

https://youtu.be/p9L7YFqHGk4?si=caoYhvQlJhyTJV50

https://youtu.be/X7guekGZM20?si=O_9xZt_zyBCQl3LN

https://youtu.be/b1RavPr_878?si=NXcUcJHVDYh-JhD4

I would personally use Zed right away if it supports .NET10 lsp and remote devcontainer.

**Comment by jtkiley** [comment]
@KyleBarton Here's one on creating a devcontainer using VS Code's interface, with a couple of common failure points and frictions noted. Happy to make more and/or answer questions.

Fair warning: This is a hit record and send it kind of video, so set expectations accordingly. 😬

https://github.com/user-attachments/assets/ba7ba5c3-492f-45e2-b0bd-0e5d423d3447

**Comment by jtkiley** [comment]
If you want to see an in situ example, I use a devcontainer for a Python course I teach to other researchers.

https://github.com/jtkiley/carma_python

This one has environment variables (passed through from secrets on Codespaces or host environment variables on your own computer) and a bit more customization via script called from the `devcontainer.json`.

I set up most of my own projects a bit differently now, but there are a couple of devcontainer-feature cleanup items I'd like to see on the uv feature before I move this course from pip to uv. It's boilerplate stuff that would be nice not to have pushed down to users (especially the new ones I teach).

**Comment by KyleBarton** [comment]
@jtkiley thanks so much for that walkthrough - super helpful! And thanks for the resources @valentin-p


It sounds like it's maybe less common than I thought that folks would package their own custom devcontainer image - but rather just use Dockerfile/docker-compose.yaml to augment from an existing `mcr.microsoft.com/devcontainers/` image.

**Comment by lucasyvas** [comment]
> [@jtkiley](https://github.com/jtkiley) thanks so much for that walkthrough - super helpful! And thanks for the resources [@valentin-p](https://github.com/valentin-p)
> 
> It sounds like it's maybe less common than I thought that folks would package their own custom devcontainer image - but rather just use Dockerfile/docker-compose.yaml to augment from an existing `mcr.microsoft.com/devcontainers/` image.

Not a conclusion I would reach. I build all my dev containers off debian slim - I think anyone that really cares about tracking their image dependencies is doing it from a neutral base image (one added benefit is if you rely on build stages for your project, since you can control all the layering). I have a dockerfile in one of my repositories that drives the dev container build by referencing it in devcontainer.json.

Edit: To elaborate further, my `docker-compose.yml` includes all the dependency containers for my service. I then "launch" my dev container in VS code (defined in devcontainer.json and its own docker file). It builds it then joins with the other already running containers. So, I _do_ heavily rely on VS Code installing its appropriate components once my devcontainer comes up since I'm not using one of Microsoft's images.

<details>
<summary>Sample Files</summary>


devcontainer.json

```json
{
  "dockerFile": "Dockerfile",
  "initializeCommand": "make network-create",
  "runArgs": ["--init", "--name=bitcloak_devcontainer", "--network=bitcloak"],
  "workspaceMount": "source=${localWorkspaceFolder},target=/bitcloak,type=bind",
  "workspaceFolder": "/bitcloak",
  "forwardPorts": [3000, 9999],
  "postStartCommand": "cargo install-update -a; psql $DATABASE_URL -c \"ALTER USER ${BITCLOAK_SERVER_PG_USERNAME} WITH PASSWORD '${BITCLOAK_SERVER_PG_PASSWORD}';\" || 0",
  "containerEnv": {
    "DATABASE_URL": "postgres://postgres:postgres@postgres/postgres",
    "BITCLOAK_SERVER_BIND_ADDR": "0.0.0.0:3000",
    "BITCLOAK_SERVER_PG_HOST": "postgres",
    "BITCLOAK_SERVER_PG_PORT": "5432",
    "BITCLOAK_SERVER_PG_USERNAME": "bitcloak_server",
    "BITCLOAK_SERVER_PG_PASSWORD": "bitcloak_server"
  },
  "customizations": {
    "vscode": {
      "extensions": [
        "42crunch.vscode-openapi",
        "esbenp.prettier-vscode",
        "github.vscode-github-actions",
        "github.vscode-pull-request-github",
        "fill-labs.dependi",
        "ms-azuretools.vscode-docker",
        "ms-vscode.live-server",
        "redhat.vscode-yaml",
        "rust-lang.rust-analyzer",
        "tamasfe.even-better-toml",
        "timonwong.shellcheck",
        "vadimcn.vscode-lldb"
      ]
    }
  }
}
```

Dockerfile

```
FROM debian:trixie-slim

ARG DEBIAN_FRONTEND=noninteractive

ARG RUST_VERSION=1.91.0
ARG NVM_VERSION=0.40.3
ARG NODE_VERSION=24.11.0
ARG POSTGRES_VERSION=18

ARG NVM_DIR=/usr/local/nvm
ARG NODE_HOME=$NVM_DIR/versions/node/$NODE_VERSION
ARG NODE_PATH=$NODE_HOME/lib/node_modules

ENV RUSTUP_HOME=/usr/local/rustup
ENV CARGO_HOME=/usr/local/cargo
ENV PATH=$CARGO_HOME/bin:$NODE_HOME/bin:$PATH

RUN apt update && apt upgrade -y && apt install -y \
  ca-certificates \
  cmake \
  curl \
  gcc \
  git \
  gnupg \
  jq \
  libc6-dev \
  libssl-dev \
  locales \
  lsb-release \
  moreutils \
  pkg-config \
  postgresql-common \
  sudo

RUN echo "en_US.UTF-8 UTF-8" > /etc/locale.gen && locale-gen

RUN echo | /usr/share/postgresql-common/pgdg/apt.postgresql.org.sh && \
  apt install -y postgresql-client-$POSTGRES_VERSION

RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --default-toolchain $RUST_VERSION && \
  cargo install cargo-update && \
  cargo install sqlx-cli --no-default-features --features native-tls,postgres

RUN /bin/bash -c "mkdir -p $NVM_DIR && \
  curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v$NVM_VERSION/install.sh | bash && \
  source $NVM_DIR/nvm.sh && \
  nvm install v$NODE_VERSION"

WORKDIR /bitcloak

CMD ["sh", "-c", "eval ${COMMAND}"]
```
</details>

**Comment by ryanhiebert** [comment]
In my experience the Microsoft ones are easy and a good base, but often fall short. I do try to use features rather than a dockerfile whenever I can, though.

**Comment by ggjulio** [comment]
I wouldn't reach the conclusion that docker compose is widely used either. 

IMHO the only valid usecase of compose was back then when the `docker in docker` feature wasn't working everywhere.

At work we usually have a Dockerfile that extends a  devcontainer debian image. We add the dind feature and done.

Then git repos may have compose files that are used the same way we would in a VM. (Thus no need to rebuild when change are made in compose files)
`kind` also works nicely.

We did this because first it's easier to think a devcontainer as a "declarative vm"/monolith defined in a simple Dockerfile,  rather than a bunch of opaque containers that add cognitive complexity, particularly during onboarding. Second because not all tools support compose. (Coder/envbuilder)

Globally it is more user friendly because we can work in devcontainers the exact way we would in a VM. Almost forgetting we're in a container.

But anyway the official devcontainer cli implem support compose so it's not something that is to be implemented in zed, it will simply work.

**Comment by ggjulio** [comment]
> In my experience the Microsoft ones are easy and a good base, but often fall short. I do try to use features rather than a dockerfile whenever I can, though.

Features takes so long to install because they are installed sequentially. I tend to avoid them.
Unfortunatly they are not even cached by the cli when using buildkit's  `--cache-to` option.

**Comment by FTWynn** [comment]
I get a lot of personal mileage out of customizing Anthropic's devcontainer recommended setup for Claude Code.

https://code.claude.com/docs/en/devcontainer

Most of my interaction in VSCode is in getting a devcontainer launched... and seeing the failure logs when any tweaks I made mean the image doesn't launch anymore. 🤦

**Comment by ggjulio** [comment]
@FTWynn FWIW  I use the devcontainer cli to iterate faster. I have a small wrapper script on top of it. (The script validate the build for arm and amd)

And then only once it build I ask vscode to rebuild and reopen.

We can even build the devcontainer inside the devcontainer if `docker in docker` feature is installed. (So it works on GitHub codespaces, gitpod/ona, etc)

**Comment by fazo96** [comment]
@KyleBarton in my team we use devcontainer in Cursor/VSCode to standardize the development environment across the whole team, which has people on Windows, Mac and Linux. We have a `.devcontainer/devcontainer.json` with a base microsoft image, and a `"features"` section to install lots of different development tools (ruby, nodejs, python, AWS CLI, EB CLI, Terraform and more) and postgresql.

Then we have `postCreateCommand` and `postStartCommand` to automatically run scripts that will set up the DB within the container and auto start it when the container starts.

The point here is that a new developer on the team just has to open the repo in VSCode/Cursor then the IDE prompts to reopen the folder in the devcontainer. They accept the prompt, and the entire development environment is automatically set up by the scripts.

Then, the entire IDE backend will run inside the container (including the terminal) and ports are automatically forwarded to the host so you can still access the web app running in the devcontainer using your host's web browser with `localhost`.

I am looking forward to trying Zed as an alternative to Cursor/VSCode for my team but being able to use devcontainer like this is a hard requirement

**Comment by sofianegargouri** [comment]
@KyleBarton I have the same usage as @fazo96. Dev Containers are really handy to handle multi-platform development and for me it's a gamechanger regarding junior developers and their environment setup. They can destroy anything in it safely and just have to rebuild the container to get started again.

I know a few folks in my team are waiting for this to make the full switch to Zed (including me)

**Comment by gintsmurans** [comment]
We also use debian base image in Dockerfile, then multi stage builds, and then docker compose to run them in both development and production. Its basically the same environment in both dev and prod, only that development stage adds few extras. And then we have devcontainer.json which, when opened in vscode, installs all the extensions project needs. This allows all devs to have the exact same environment regardless of OS they are using and also have same code formatting extensions that are also only installed inside container.

**Comment by baldwindavid** [comment]
Our team uses devcontainers for cross-platform development that just works in vscode. The `devcontainer.json` is very light and mostly just delegates to a dockerfile. I do a separate build that merges my own additional dockerfile and devcontainer settings and builds via dev container CLI. I really just want to easily open an already built devcontainer in Zed. As it is, I'm accessing the built container in Zed via SSH (when it's not broken due to my own unstable scripts).

**Comment by FTWynn** [comment]
One other important note about my VSCode usage: the connection of extensions and such _into the container_, rather than the host machine, is pretty essential. Claude Code, for example, is not installed at all on my host. That may have some significant impact on Zed AI's implementation.

**Comment by lucasyvas** [comment]
Another nice thing that works is that when I launch my container, VS Code facilitates the synchronization of my host git config to the container automatically, so I can push and pull to/from the repository in the same way that I initially cloned it down onto the host.

**Comment by ryanhiebert** [comment]
And the SSH agent forwarding makes it all seamless.

**Comment by aidandj** [comment]
> One other important note about my VSCode usage: the connection of extensions and such _into the container_, rather than the host machine, is pretty essential. Claude Code, for example, is not installed at all on my host. That may have some significant impact on Zed AI's implementation.

Extensions can run both in the container or on the local machine, depending on the extension and configuration.

**Comment by fkromer** [comment]
As temporary workaround this could be interesting: https://dev.to/ale_annini/lightning-fast-development-with-zed-and-dev-containers-1nbd

**Comment by fkromer** [comment]
Support for devcontainers would make Zed play well with Bluefin OS https://docs.projectbluefin.io/devcontainers/ (on Bluefin it‘s not straight forward to use devbox as devcontainer alternative).

**Comment by FY0u11** [comment]
Any progress on this? ETA? Ty

**Comment by prponkshe** [comment]
Till zed has it's own devcontainer extension up and running, I've been using the remote development module to ssh into the containers, by exposing a port. I run sshd via the features property in the devcontainer.json, so I don't need to make any changes to my Dockerfile. And a postStartCommand to start the ssh service. 

Works like a charm. I've made a bash function that tracks all container-port mappings, and auto-assigns ports in generally unused ranges. 

Only caveat is it doesn't work for devcontainers that use docker compose to spin up multiple containers at a time.

**Comment by versecafe** [comment]
Anything that auto aliases docker should work, from my testing orbstack for example works perfectly without changes as they alias docker by default

**Comment by kyleburke-meq** [comment]
Any chance this can support multiple devcontainers? For example I have many containers in in my .devconatiner directory. Ideally zed would prompt for which I want to use:

<img width="317" height="354" alt="Image" src="https://github.com/user-attachments/assets/e893c3e2-9347-4827-a6eb-f1b9bffddaf2" />

This feature is supported in vscode:
https://code.visualstudio.com/remote/advancedcontainers/connect-multiple-containers

**Comment by imbev** [comment]
IMO the user should be prompted to allow SSH/GPG forwarding, per-project or as a saved preference. The user should also have the option to mount the ~/.ssh directory instead of the SSH Agent, regardless of the project's devcontainer.json

**Comment by SinBirb** [comment]
[Environment variables](https://containers.dev/implementors/json_reference/#variables-in-devcontainerjson) like `localWorkspaceFolder` and `containerWorkspaceFolder` are not yet implemented, correct? At least on an existing VSCode devcontainer they are not populated in a Zed terminal.

**Comment by neoSnakex34** [comment]
@KyleBarton I love how fast new features are implemented and improved, it means a lot to zed users!

**Comment by tib-rijo** [comment]
@neoSnakex34 

> Would be nice to implement an option to close connection from the devcontianer

I'd figure this is implemented with closing the Zed window. Can you confirm that you can disconnect Zed from the container by closing Zed? A new Zed instance / window will not be connected (anymore).

Interesting here would be for me, if Zed can also manage the Dev Container lifecycle for me when closing itself, such as optionally stopping / removing the Dev Container. Right now it leaves it untouched and running.

**Comment by almereyda** [comment]
> On a related note, Git inside the devcontainer does not have access to my credentials, so I cannot pull or push.

Is there a way to keep credentials in `secretservice` and map that into the container, very much like FlatPak does? Also thinking of [Docker/Podman credential helpers](https://linuskarlsson.se/blog/podman-credential-helpers/) for OCI Distribution registries or `gh`/`glab`/`berg` CLIs.

**Comment by mathroc** [comment]
the Zed debugger works just fine in devcontainers here, that's most likely a configuration issue, either in Zed, or in PHP. you can run `xdebug_info()` and look at the "Step Debugger" section to see if it succeeds or why xdebug didn't to connect the Zed debugger

**Comment by wjjsn** [comment]
still connot to use it

<img width="823" height="492" alt="Image" src="https://github.com/user-attachments/assets/b69afbc6-f54a-4f45-88b1-d4609cc90b87" />
